src/contracts/BCoWPool.sol

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.25;
+
+/*
+
+Coded for Balancer and CoW Swap with ♥ by
+
+░██╗░░░░░░░██╗░█████╗░███╗░░██╗██████╗░███████╗██████╗░██╗░░░░░░█████╗░███╗░░██╗██████╗░
+░██║░░██╗░░██║██╔══██╗████╗░██║██╔══██╗██╔════╝██╔══██╗██║░░░░░██╔══██╗████╗░██║██╔══██╗
+░╚██╗████╗██╔╝██║░░██║██╔██╗██║██║░░██║█████╗░░██████╔╝██║░░░░░███████║██╔██╗██║██║░░██║
+░░████╔═████║░██║░░██║██║╚████║██║░░██║██╔══╝░░██╔══██╗██║░░░░░██╔══██║██║╚████║██║░░██║
+░░╚██╔╝░╚██╔╝░╚█████╔╝██║░╚███║██████╔╝███████╗██║░░██║███████╗██║░░██║██║░╚███║██████╔╝
+░░░╚═╝░░░╚═╝░░░╚════╝░╚═╝░░╚══╝╚═════╝░╚══════╝╚═╝░░╚═╝╚══════╝╚═╝░░╚═╝╚═╝░░╚══╝╚═════╝░
+
+https://defi.sucks
+
+*/
+
+import {BCoWConst} from './BCoWConst.sol';
+import {BPool} from './BPool.sol';
+import {GPv2Order} from '@cowprotocol/libraries/GPv2Order.sol';
+import {IERC1271} from '@openzeppelin/contracts/interfaces/IERC1271.sol';
+import {IERC20} from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+
+import {IBCoWFactory} from 'interfaces/IBCoWFactory.sol';
+import {IBCoWPool} from 'interfaces/IBCoWPool.sol';
+import {ISettlement} from 'interfaces/ISettlement.sol';
+
+/**
+ * @title BCoWPool
+ * @notice Pool contract that holds tokens, allows to swap, add and remove liquidity.
+ * @dev Inherits BPool contract functionalities, and can trade on CoW Swap Protocol.
+ */
+contract BCoWPool is IERC1271, IBCoWPool, BPool, BCoWConst {
+  using GPv2Order for GPv2Order.Data;
+
+  /// @inheritdoc IBCoWPool
+  address public immutable VAULT_RELAYER;
+
+  /// @inheritdoc IBCoWPool
+  bytes32 public immutable SOLUTION_SETTLER_DOMAIN_SEPARATOR;
+
+  /// @inheritdoc IBCoWPool
+  ISettlement public immutable SOLUTION_SETTLER;
+
+  /// @inheritdoc IBCoWPool
+  bytes32 public immutable APP_DATA;
+
+  constructor(address _cowSolutionSettler, bytes32 _appData) BPool() {
+    SOLUTION_SETTLER = ISettlement(_cowSolutionSettler);
+    SOLUTION_SETTLER_DOMAIN_SEPARATOR = ISettlement(_cowSolutionSettler).domainSeparator();
+    VAULT_RELAYER = ISettlement(_cowSolutionSettler).vaultRelayer();
+    APP_DATA = _appData;
+  }
+
+  /// @inheritdoc IBCoWPool
+  function commit(bytes32 orderHash) external {
+    if (msg.sender != address(SOLUTION_SETTLER)) {
+      revert CommitOutsideOfSettlement();
+    }
+    if (_getLock() != _MUTEX_FREE) {
+      revert BCoWPool_CommitmentAlreadySet();
+    }
```

Comment:
could be replaced by `_viewlock_` 🤔

Reply:
we would get another, less descriptive error message. Other than that, I agree